From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Markus Rudy <mr@edgeless.systems>
Date: Fri, 6 Dec 2024 17:14:57 +0100
Subject: [PATCH] genpolicy: support guest hooks

Some custom images come with guest hooks that are required to set up
VM-level resources for containers, such as GPU support.

This commit adds a new setting, `allowed_guest_hooks`, which controls
what paths may be set for hooks.

Signed-off-by: Markus Rudy <mr@edgeless.systems>
---
 src/tools/genpolicy/genpolicy-settings.json |  3 ++-
 src/tools/genpolicy/rules.rego              | 17 ++++++++++++++++-
 src/tools/genpolicy/src/policy.rs           |  2 ++
 3 files changed, 20 insertions(+), 2 deletions(-)

diff --git a/src/tools/genpolicy/genpolicy-settings.json b/src/tools/genpolicy/genpolicy-settings.json
index 21a9d6a9005765ac2852fd82d5f2aa1e6237b2bf..d60fd607c08d6b403c0bf386d9b41535e54428ff 100644
--- a/src/tools/genpolicy/genpolicy-settings.json
+++ b/src/tools/genpolicy/genpolicy-settings.json
@@ -247,7 +247,8 @@
                 "mount_point": "/run/kata-containers/sandbox/shm",
                 "fs_group": null
             }
-        ]
+        ],
+        "allowed_guest_hooks": ["/usr/share/oci/hooks"]
     },
     "common": {
         "cpath": "/run/kata-containers",
diff --git a/src/tools/genpolicy/rules.rego b/src/tools/genpolicy/rules.rego
index a0387132fdd59c03ab54502432d8a97105454a8b..4d3385010491b00cdf12fc20a592eb8f12c3d74d 100644
--- a/src/tools/genpolicy/rules.rego
+++ b/src/tools/genpolicy/rules.rego
@@ -1413,7 +1413,7 @@ CopyFileRequest {
 
 CreateSandboxRequest {
     print("CreateSandboxRequest: input.guest_hook_path =", input.guest_hook_path)
-    count(input.guest_hook_path) == 0
+    allow_hook(input.guest_hook_path)
 
     print("CreateSandboxRequest: input.kernel_modules =", input.kernel_modules)
     count(input.kernel_modules) == 0
@@ -1424,6 +1424,21 @@ CreateSandboxRequest {
     allow_sandbox_storages(input.storages)
 }
 
+allow_hook(i_hook) {
+    print("allow_hook 1: start")
+    count(i_hook) == 0
+    print("allow_hook 1: true")
+}
+
+allow_hook(i_hook) {
+    print("allow_hook 2: start")
+    p_hooks := policy_data.sandbox.allowed_guest_hooks
+    print("allow_hook 2: p_hooks =", p_hooks)
+    print("allow_hook 2: i_hook =", i_hook)
+    i_hook in p_hooks
+    print("allow_hook 2: true")
+}
+
 allow_exec(p_container, i_process) {
     print("allow_exec: start")
 
diff --git a/src/tools/genpolicy/src/policy.rs b/src/tools/genpolicy/src/policy.rs
index de101545bd68d2973cc89254d33b59cc4ef28245..a927dc8090a7b65e60b9d289a4b5c7b72c65a740 100644
--- a/src/tools/genpolicy/src/policy.rs
+++ b/src/tools/genpolicy/src/policy.rs
@@ -439,6 +439,8 @@ pub struct ClusterConfig {
 pub struct SandboxData {
     /// Expected value of the CreateSandboxRequest storages field.
     pub storages: Vec<agent::Storage>,
+    /// Guest hook paths that are allowed to be set in all CreateContainerRequests.
+    pub allowed_guest_hooks: Vec<String>,
 }
 
 enum K8sEnvFromSource {
